modules/gapi/src/compiler/passes/pattern_matching.cpp

```diff
@@ -0,0 +1,371 @@
+#include <unordered_set>
+
+#include "pattern_matching.hpp"
+
+cv::gapi::SubgraphMatch cv::gapi::findMatches(cv::gimpl::GModel::Graph patternGraph, cv::gimpl::GModel::Graph compGraph) {
+    using Graph = cv::gimpl::GModel::Graph;
+    using Metadata = typename Graph::MetadataT;
+    using VisitedMatchings = std::list<std::pair<ade::NodeHandle, ade::NodeHandle>>;
+
+    //TODO: Possibly, we may add N^2 check whether this graph may match or not at all.
+    //      Check that all pattern OP nodes exist in computational graph.
+
+    std::unordered_set<ade::NodeHandle, ade::HandleHasher<ade::Node>> firstPatternOpNodes;
+    std::unordered_set<ade::NodeHandle, ade::HandleHasher<ade::Node>> lastPatternOpNodes;
+
+    auto firstPatternDataNodes = patternGraph.metadata().get<cv::gimpl::Protocol>().in_nhs;
+    auto lastPatternDataNodes = patternGraph.metadata().get<cv::gimpl::Protocol>().out_nhs;
+
+    for (auto node : firstPatternDataNodes) {
+        auto opNodes = node->outNodes();
+        firstPatternOpNodes.insert(opNodes.begin(), opNodes.end());
+    }
+
+    for (auto node : lastPatternDataNodes) {
+        auto opNodes = node->inNodes();
+        lastPatternOpNodes.insert(opNodes.begin(), opNodes.end());
+    }
+
+    // For visited matchings to work correctly it is required that output nodes from some node found in both graphs were in the same order!
+    // And that outNodes array contains these nodes right in the order that they are presented in graph (left-to-right nodes-exact).
+    // Else graphs were not assumed to be equal.
+    VisitedMatchings matchedVisitedNodes;
+
+    auto dataNodesComparator = [](std::pair<const ade::NodeHandle, std::vector<std::size_t>> first, Metadata firstMetadata, std::pair<const ade::NodeHandle, std::vector<std::size_t>> second, Metadata secondMetadata) {
+        if (secondMetadata.get<cv::gimpl::NodeType>().t != cv::gimpl::NodeType::DATA) {
+            //TODO: FIX ASAP
+            throw std::logic_error("NodeType of passed node as second argument shall be NodeType::DATA!");
+        }
+
+        if (firstMetadata.get<cv::gimpl::Data>().shape != secondMetadata.get<cv::gimpl::Data>().shape) {
+            return false;
+        }
+
+        //todo
+        std::sort(first.second.begin(), first.second.end());
+        std::sort(second.second.begin(), second.second.end());
+        if (first.second != second.second) {
+            return false;
+        }
+
+        auto firstOutputNodes = first.first->outNodes();
+        auto secondOutputNodes = second.first->outNodes();
+
+        if (firstOutputNodes.size() != secondOutputNodes.size()) {
+            return false;
+        }
+
+        return true;
+    };
+
+    auto opNodesComparator = [&matchedVisitedNodes](std::pair<const ade::NodeHandle, std::vector<std::size_t>> first, Metadata firstMetadata, std::pair<const ade::NodeHandle, std::vector<std::size_t>> second, Metadata secondMetadata, bool& isAlreadyVisited) {
```

Comment:
Please reduce code lines length. It's hard to review. Please divide on several lines. This is also for the whole ammount of code for review.

Reply:
Shall be done after refactoring

Reply:
Done